energy-integration/farm-boosted-yields/src/custom_reward_logic.rs

```diff
@@ -0,0 +1,248 @@
+        }
+
+        last_collect_week_mapper.set(last_collect_week);
+    }
+
+    fn take_reward_slice(&self, full_reward: BigUint) -> SplitReward<Self::Api> {
+        let percentage = self.boosted_yields_rewards_percentage().get();
+        if percentage == 0 {
+            return SplitReward::new(full_reward, BigUint::zero());
+        }
+
+        let boosted_yields_cut = &full_reward * percentage / MAX_PERCENT;
+        let base_farm_amount = if boosted_yields_cut > 0 {
+            let current_week = self.get_current_week();
+            self.accumulated_rewards_for_week(current_week)
+                .update(|accumulated_rewards| {
+                    *accumulated_rewards += &boosted_yields_cut;
+                });
+
+            &full_reward - &boosted_yields_cut
+        } else {
+            full_reward
+        };
+
+        SplitReward::new(base_farm_amount, boosted_yields_cut)
+    }
+
+    fn calculate_user_boosted_rewards(&self, args: CalculateRewardsArgs<Self::Api>) -> BigUint {
+        let max_rewards =
+            &args.factors.max_rewards_factor * args.weekly_reward_amount * args.user_farm_amount
+                / args.farm_supply_for_week;
+
+        // computed user rewards = total_boosted_rewards *
+        // (energy_const * user_energy / total_energy + farm_const * user_farm / total_farm) /
+        // (energy_const + farm_const)
+        let boosted_rewards_by_energy = args.weekly_reward_amount
+            * &args.factors.user_rewards_energy_const
+            * args.energy_amount
+            / args.total_energy;
+        let boosted_rewards_by_tokens = args.weekly_reward_amount
+            * &args.factors.user_rewards_farm_const
+            * args.user_farm_amount
+            / args.farm_supply_for_week;
+        let constants_base =
+            &args.factors.user_rewards_energy_const + &args.factors.user_rewards_farm_const;
+        let boosted_reward_amount =
+            (boosted_rewards_by_energy + boosted_rewards_by_tokens) / constants_base;
+
+        core::cmp::min(max_rewards, boosted_reward_amount)
+    }
+
+    fn limit_boosted_rewards_by_claim_time(
+        &self,
+        user_reward: BigUint,
+        week_timestamps: &WeekTimestamps,
+        claim_progress: &ClaimProgress<Self::Api>,
+    ) -> BigUint {
+        if !(claim_progress.enter_timestamp >= week_timestamps.start
+            && claim_progress.enter_timestamp < week_timestamps.end)
+        {
+            return user_reward;
+        }
+
+        // Example: user entered at 25% of week, so give only 75% of rewards
+        let enter_timestamp_percent_of_week = linear_interpolation::<Self::Api, _>(
+            week_timestamps.start as u128,
+            week_timestamps.end as u128,
+            claim_progress.enter_timestamp as u128,
+            0,
+            MAX_PERCENT as u128,
+        );
+        let percent_leftover = MAX_PERCENT as u128 - enter_timestamp_percent_of_week;
```

Comment:
Why all of this u128 converting?

Reply:
I don't think we could overflow. We have a maximum number of 6_048_000_000 (max number of seconds in a week * MAX_PERCENT), which is under u64 max limit.